serialization/src/extras/non_empty_vec.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use serialization_core::{Decode, Encode};
+
+/// A wrapper that saves the extra byte coming from Option by assuming the Vec can never be empty
+///
+/// Description:
+/// - If the Vec has data, it encodes to just the Vec, the Option is omitted
+/// - If the Vec has no data, it encodes to None
+/// - Some(vec![]) and None are equivalent when encoded, but when decoded result in None
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct DataOrNoVec<T>(Option<Vec<T>>);
+
+impl<T> AsRef<Option<Vec<T>>> for DataOrNoVec<T> {
+    fn as_ref(&self) -> &Option<Vec<T>> {
+        &self.0
+    }
+}
+
+impl<T> From<Option<Vec<T>>> for DataOrNoVec<T> {
+    fn from(val: Option<Vec<T>>) -> Self {
+        match val {
+            Some(ref v) => {
+                if v.is_empty() {
+                    DataOrNoVec(None)
+                } else {
+                    DataOrNoVec(val)
+                }
+            }
+            None => DataOrNoVec(None),
+        }
+    }
+}
+
+impl<T> From<DataOrNoVec<T>> for Option<Vec<T>> {
+    fn from(val: DataOrNoVec<T>) -> Self {
+        val.0
+    }
+}
+
+impl<U: Encode> Encode for DataOrNoVec<U> {
+    fn encode_to<T: serialization_core::Output + ?Sized>(&self, dest: &mut T) {
+        match &self.0 {
+            Some(v) => v.encode_to(dest),
+            None => Vec::<U>::new().encode_to(dest),
+        }
+    }
+
+    fn encoded_size(&self) -> usize {
+        match &self.0 {
+            Some(v) => v.encoded_size(),
+            None => Vec::encoded_size(&Vec::<U>::new()),
+        }
+    }
+}
+
+impl<U: Decode> Decode for DataOrNoVec<U> {
+    fn decode<I: serialization_core::Input>(
+        input: &mut I,
+    ) -> Result<Self, serialization_core::Error> {
+        let v = Vec::decode(input)?;
+        if v.is_empty() {
+            Ok(DataOrNoVec(None))
+        } else {
+            Ok(DataOrNoVec(Some(v)))
+        }
```

Comment:
I think this can now delegate to the `from` method:
```rust
Vec::decode(input).map(Self::from)
```

Reply:
Not exactly, but I did the closest to that.

Reply:
Even though it's a little less efficient, btw... because we have to match again in From